lib/runtime/life/instance.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+package life
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/genesis"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	log "github.com/ChainSafe/log15"
+	"github.com/perlin-network/life/exec"
+)
+
+// Name represents the name of the interpreter
+const Name = "life"
+
+// Check that runtime interfaces are satisfied
+var (
+	_      runtime.Instance = (*Instance)(nil)
+	logger                  = log.New("pkg", "runtime", "module", "perlin/life")
+	ctx    *runtime.Context
+)
+
+// Config represents a life configuration
+type Config struct {
+	runtime.InstanceConfig
+	Resolver exec.ImportResolver
+}
+
+// Instance represents a v0.8 runtime life instance
+type Instance struct {
+	vm      *exec.VirtualMachine
+	mu      sync.Mutex
+	version *runtime.VersionAPI
+}
+
+// NewRuntimeFromGenesis creates a runtime instance from the genesis data
+func NewRuntimeFromGenesis(g *genesis.Genesis, cfg *Config) (runtime.Instance, error) { // TODO: simplify, get :code from storage
+	codeStr := g.GenesisFields().Raw["top"][common.BytesToHex(common.CodeKey)]
+	if codeStr == "" {
+		return nil, fmt.Errorf("cannot find :code in genesis")
+	}
+
+	code := common.MustHexToBytes(codeStr)
+	cfg.Resolver = new(Resolver)
+	return NewInstance(code, cfg)
+}
+
+// NewInstance ...
+func NewInstance(code []byte, cfg *Config) (runtime.Instance, error) {
+	if len(code) == 0 {
+		return nil, errors.New("code is empty")
+	}
+
+	// if cfg.LogLvl set to < 0, then don't change package log level
+	if cfg.LogLvl >= 0 {
+		h := log.StreamHandler(os.Stdout, log.TerminalFormat())
+		h = log.CallerFileHandler(h)
+		logger.SetHandler(log.LvlFilterHandler(cfg.LogLvl, h))
+	}
+
+	instance, err := exec.NewVirtualMachine(code, exec.VMConfig{
```

Comment:
Is the formatting/indentation correct here?

Reply:
yeah, i just indented the config while it's a param, i'll separate it out